lightly/models/modules/masked_causal_vision_transformer.py

```diff
@@ -0,0 +1,247 @@
+from typing import Callable, Literal, Optional, Tuple, Type, Union
+
+import torch
+import torch.nn.functional as F
+from timm.layers import LayerType, Mlp, PatchEmbed
+from timm.models import _manipulate
+from timm.models.vision_transformer import Attention, Block, VisionTransformer
+from torch import Tensor, jit
+from torch.nn import GELU, LayerNorm, Module
+
+
+# Type ignore because superclass has Any types.
+class MaskedCausalAttention(Attention):  # type: ignore[misc]
+    """Identical to timm.models.vision_transformer.Attention, but supports causal
+    attention with masking.
+
+    The implementation is based on AIM [0].
+
+    - [0]: AIM, 2024, https://arxiv.org/abs/2401.08541
+    """
+
+    def forward(self, x: Tensor, mask: Optional[Tensor] = None) -> Tensor:
+        """Forward pass of the attention layer.
+
+        Args:
+            x:
+                Input tensor of shape (batch_size, sequence_length, channels).
+            mask:
+                Mask of shape (batch_size, sequence_length) indicating which tokens
+                should be masked. Tokens where the mask is True will only be used for
+                causal attention, while unmasked tokens are used for bidirectional
+                attention. If the mask is None, all tokens are used for bidirectional
+                attention.
+        """
+        B, N, C = x.shape
+
+        # Only apply causal attention if mask is not None. This is a bit hacky, but it
+        # allows us to use bidirectional instead of causal attention during evaluation
+        # and fine-tuning.
+        attn_mask = None
+        if mask is not None:
+            attn_mask = x.new_ones(size=(B, N, N), dtype=torch.bool).tril(diagonal=0)
+            # mask has shape (B, N)
+            mask = (~mask).unsqueeze(1).expand(B, N, N).bool()
+            attn_mask = torch.logical_or(attn_mask, mask)
+            attn_mask = attn_mask.unsqueeze(1)  # (B, 1, N, N)
+
```

Comment:
I think here it would make sense to have a separate method that handles the masking (with a comment also pointing to the code of the paper) and then use the `forward` method of TIMM's [`Attention`](https://github.com/huggingface/pytorch-image-models/blob/53a48883280743f5c878897bdfd3d471f1c191e4/timm/models/vision_transformer.py#L84).

Reply:
Good point, a separate method definitely makes sense. But I cannot use timm's attention as it doesn't pass the attention mask to the attention calculation: https://github.com/huggingface/pytorch-image-models/blob/53a48883280743f5c878897bdfd3d471f1c191e4/timm/models/vision_transformer.py#L92